File: Saphal Book.app/Contents/Resources/app/chartered_book/modules/banking.py
"""
Cash, bank and reconciliation.

A bank reconciliation answers one question: our books say the account holds
this much, the bank statement says something else, and the difference has to be
explained item by item rather than written off.

The working here is the standard one:

    Balance as per our books
      less  deposits we have recorded that the bank has not yet credited
      plus  cheques we have issued that have not yet been presented
      ----
    Balance that the bank statement should show

Nothing is altered in the books by reconciling. Ticking an entry as cleared
only records the date the bank dealt with it.
"""

from ..core import audit, db, money
from . import masters, reports


class BankingError(Exception):
    """Raised when a banking action cannot be carried out."""


CASH_GROUP = "1250"
BANK_GROUP = "1260"
OVERDRAFT_GROUP = "2220"


def accounts(conn, as_at_ad=None, include_inactive=False):
    """Every cash, bank and overdraft ledger with its balance."""
    rows = conn.execute(
        """SELECT a.*, g.code AS group_code, g.name AS group_name
           FROM accounts a JOIN account_groups g ON g.id = a.group_id
           WHERE a.account_kind IN ('cash', 'bank')
           %s
           ORDER BY a.account_kind DESC, a.code""" % ("" if include_inactive else "AND a.active = 1")
    ).fetchall()
    balances = reports.balances_as_at(conn, as_at_ad) if as_at_ad else {}
    out = []
    for row in rows:
        record = dict(row)
        record["balance"] = balances.get(row["id"], row["opening_paisa"])
        record["uncleared"] = uncleared_total(conn, row["id"], as_at_ad) if as_at_ad else 0
        out.append(record)
    return out


def create_account(conn, username, name, kind="bank", **fields):
    """
    Add a bank account, a cash box or an overdraft.

    A bank account goes under Bank Balances, cash under Cash in Hand, and an
    overdraft under Short Term Borrowings because it is money owed, not held.
    """
    if kind not in ("bank", "cash", "overdraft"):
        raise BankingError("Choose bank, cash or overdraft.")
    group_code = {"bank": BANK_GROUP, "cash": CASH_GROUP, "overdraft": OVERDRAFT_GROUP}[kind]
    group = masters.group_by_code(conn, group_code)
    if group is None:
        raise BankingError("The %s group is missing from the chart of accounts." % group_code)
    return masters.create_account(
        conn, username, name, group["id"],
        account_kind="cash" if kind == "cash" else "bank",
        name_np=fields.get("name_np", ""),
        bank_name=fields.get("bank_name", ""),
        bank_account_no=fields.get("bank_account_no", ""),
        bank_branch=fields.get("bank_branch", ""),
        reconcilable=0 if kind == "cash" else 1,
        opening=fields.get("opening") or 0,
        opening_side=fields.get("opening_side", "cr" if kind == "overdraft" else "dr"),
        notes=fields.get("notes", ""))


def entries_for(conn, account_id, upto_ad, only_uncleared=False):
    """Every posting in one cash or bank ledger, with its clearance state."""
    sql = """SELECT e.id AS entry_id, e.dr_paisa, e.cr_paisa, e.narration, e.cleared_ad,
                    e.instrument_no, e.instrument_date_ad,
                    v.id AS voucher_id, v.number, v.date_ad, v.date_bs, v.voucher_type,
                    v.narration AS voucher_narration, v.reference_no,
                    p.name AS party_name,
                    (SELECT GROUP_CONCAT(a2.name, ', ')
                       FROM voucher_entries e2 JOIN accounts a2 ON a2.id = e2.account_id
                      WHERE e2.voucher_id = v.id AND e2.account_id <> ?) AS contra
             FROM voucher_entries e
             JOIN vouchers v ON v.id = e.voucher_id
             LEFT JOIN parties p ON p.id = v.party_id
             WHERE e.account_id = ? AND v.status = 'posted' AND v.date_ad <= ?"""
    args = [account_id, account_id, upto_ad]
    if only_uncleared:
        sql += " AND e.cleared_ad = ''"
    sql += " ORDER BY v.date_ad, v.id, e.line_no"
    return conn.execute(sql, args).fetchall()


def uncleared_total(conn, account_id, upto_ad):
    row = conn.execute(
        """SELECT COALESCE(SUM(e.dr_paisa), 0) AS dr, COALESCE(SUM(e.cr_paisa), 0) AS cr
           FROM voucher_entries e JOIN vouchers v ON v.id = e.voucher_id
           WHERE e.account_id = ? AND v.status = 'posted'
             AND v.date_ad <= ? AND e.cleared_ad = ''""", (account_id, upto_ad)).fetchone()
    return row["dr"] - row["cr"]


def worksheet(conn, account_id, statement_date_ad):
    """
    Everything the reconciliation screen needs.

    Returns the ledger balance, the entries with their clearance state, the two
    lists of items that explain the difference, and the bank balance those
    figures imply.
    """
    account = masters.get_account(conn, account_id)
    if account is None:
        raise BankingError("That account no longer exists.")
    if account["account_kind"] not in ("cash", "bank"):
        raise BankingError("%s is not a cash or bank account." % account["name"])

    balances = reports.balances_as_at(conn, statement_date_ad)
    book_balance = balances.get(account_id, account["opening_paisa"])

    lines = []
    uncleared_dr = uncleared_cr = 0
    for row in entries_for(conn, account_id, statement_date_ad):
        cleared = bool(row["cleared_ad"])
        if not cleared:
            uncleared_dr += row["dr_paisa"]
            uncleared_cr += row["cr_paisa"]
        lines.append({
            "entry_id": row["entry_id"],
            "voucher_id": row["voucher_id"],
            "number": row["number"],
            "voucher_type": row["voucher_type"],
            "date_ad": row["date_ad"],
            "date_bs": row["date_bs"],
            "party_name": row["party_name"] or "",
            "particulars": row["contra"] or row["narration"] or row["voucher_narration"] or "",
            "instrument_no": row["instrument_no"] or row["reference_no"] or "",
            "dr": row["dr_paisa"],
            "cr": row["cr_paisa"],
            "cleared": cleared,
            "cleared_ad": row["cleared_ad"],
        })

    # A debit in our books is money we say came in. Until the bank credits it,
    # the statement is lower than our books by that amount.
    implied_statement = book_balance - uncleared_dr + uncleared_cr

    previous = conn.execute(
        """SELECT * FROM reconciliations WHERE account_id = ? AND statement_date_ad <= ?
           ORDER BY statement_date_ad DESC, id DESC LIMIT 1""",
        (account_id, statement_date_ad)).fetchone()

    return {
        "account": dict(account),
        "statement_date_ad": statement_date_ad,
        "book_balance": book_balance,
        "uncleared_deposits": uncleared_dr,
        "uncleared_payments": uncleared_cr,
        "implied_statement_balance": implied_statement,
        "cleared_count": sum(1 for line in lines if line["cleared"]),
        "uncleared_count": sum(1 for line in lines if not line["cleared"]),
        "lines": lines,
        "last_reconciliation": dict(previous) if previous else None,
    }


def set_cleared(conn, username, entry_ids, cleared_ad, account_id=None):
    """Tick or untick entries. Passing a blank date marks them as not cleared."""
    if not entry_ids:
        return 0
    marks = ", ".join("?" for _ in entry_ids)
    conn.execute("UPDATE voucher_entries SET cleared_ad = ? WHERE id IN (%s)" % marks,
                 [cleared_ad or ""] + list(entry_ids))
    audit.log(conn, username, "bank.clear", "voucher_entries", account_id, "",
              "%d entries marked %s." % (len(entry_ids),
                                         "cleared on " + cleared_ad if cleared_ad else "not cleared"))
    return len(entry_ids)


def save_reconciliation(conn, username, account_id, statement_date_ad,
                        statement_balance, note="", complete=False):
    """Record the reconciliation so the next one can start where this left off."""
    sheet = worksheet(conn, account_id, statement_date_ad)
    statement_balance = money.to_paisa(statement_balance)
    difference = statement_balance - sheet["implied_statement_balance"]
    if complete and difference != 0:
        raise BankingError(
            "The reconciliation is out by %s. Tick the entries the bank has dealt with, "
            "or enter the missing voucher, before marking it complete."
            % money.format_money(abs(difference)))
    now = db.now_stamp()
    from ..core import nepali_date as nd
    cur = conn.execute(
        """INSERT INTO reconciliations (account_id, statement_date_ad, statement_date_bs,
                                        statement_balance, book_balance, difference,
                                        status, note, created_by, created_at, completed_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (account_id, statement_date_ad,
         nd.format_bs(nd.ad_to_bs(statement_date_ad), "numeric"),
         statement_balance, sheet["book_balance"], difference,
         "completed" if complete else "open", note, username, now,
         now if complete else ""))
    audit.log(conn, username, "bank.reconcile", "reconciliations", cur.lastrowid,
              sheet["account"]["name"],
              "Reconciled to %s at %s. Difference %s."
              % (statement_date_ad, money.format_money(statement_balance),
                 money.format_money(difference)))
    return cur.lastrowid


def history(conn, account_id=None, limit=50):
    sql = """SELECT r.*, a.name AS account_name FROM reconciliations r
             JOIN accounts a ON a.id = r.account_id"""
    args = []
    if account_id:
        sql += " WHERE r.account_id = ?"
        args.append(account_id)
    sql += " ORDER BY r.statement_date_ad DESC, r.id DESC LIMIT ?"
    args.append(limit)
    return conn.execute(sql, args).fetchall()


def cash_position(conn, as_at_ad):
    """A one line summary for the dashboard and the banking screen."""
    rows = accounts(conn, as_at_ad)
    cash = sum(r["balance"] for r in rows if r["account_kind"] == "cash")
    bank = sum(r["balance"] for r in rows if r["account_kind"] == "bank")
    return {"cash": cash, "bank": bank, "total": cash + bank, "accounts": rows}

File: chartered_book/web/static/banking.js
/* Cash, bank accounts and reconciliation.

   Reconciling never changes a figure in the books. Ticking a line only records
   that the bank has dealt with it, which is what turns an unexplained
   difference into a list of items anyone can check. */

var Banking = (function () {
  "use strict";

  var el = UI.el, api = UI.api;

  /* Cash and bank */

  App.register("banking", function (page) {
    var box = el("div");

    function load() {
      return api("/api/banking/accounts").then(function (data) {
        var cash = data.position.cash;
        var bank = data.position.bank;
        UI.clear(box);
        box.appendChild(el("div.grid.three", {}, [
          tile("Cash in hand", UI.rs(cash), "Across every cash box"),
          tile("At the bank", UI.rs(bank), "Across every account"),
          tile("Total on hand", UI.rs(data.position.total), "As at " + UI.bs(data.as_at, "long"),
               data.position.total >= 0 ? "good" : "bad")
        ]));

        var rows = data.rows.map(function (account) {
          return el("tr.clickable", { onclick: function () { Reports.openLedger(account.id); } }, [
            el("td", { text: account.code }),
            el("td", {}, [
              el("div", { text: account.name }),
              account.bank_name || account.bank_account_no
                ? el("div.muted", { style: "font-size:.75rem", text:
                    [account.bank_name, account.bank_account_no, account.bank_branch]
                      .filter(Boolean).join("  ") })
                : null
            ]),
            el("td", {}, [el("span.pill" + (account.account_kind === "cash" ? ".warn" : ".brand"),
              { text: account.account_kind })]),
            el("td.num", { text: UI.rs(account.balance) }),
            el("td.num", { text: account.uncleared ? UI.rs(account.uncleared) : "" }),
            el("td.no-print", {}, [
              account.account_kind === "bank"
                ? el("button.link-button", { text: "Reconcile", onclick: function (event) {
                    event.stopPropagation();
                    App.state.pendingRecon = account.id;
                    App.go("reconcile");
                  }})
                : null
            ])
          ]);
        });

        box.appendChild(el("div.card", { style: "margin-top:.9rem" }, [
          el("div.card-head", {}, [
            el("h2", { text: "Cash boxes and bank accounts" }),
            el("div.row", {}, [
              el("button.secondary", { text: "Add a cash box",
                onclick: function () { openAccountForm("cash"); } }),
              el("button.secondary", { text: "Add an overdraft",
                onclick: function () { openAccountForm("overdraft"); } }),
              el("button.primary", { text: "Add a bank account",
                onclick: function () { openAccountForm("bank"); } })
            ])
          ]),
          UI.table([
            "Code", "Account", "Kind",
            { label: "Balance in books", num: true },
            { label: "Not yet cleared", num: true },
            { label: "", num: false }
          ], rows, null, {
            emptyText: "No cash or bank accounts yet. Add the ones this business uses."
          })
        ]));
      });
    }

    page.appendChild(box);
    return load();
  });

  function tile(label, value, note, kind) {
    return el("div.tile" + (kind ? "." + kind : ""), {}, [
      el("div.tile-label", { text: label }),
      el("div.tile-value", { text: value }),
      note ? el("div.tile-note", { text: note }) : null
    ]);
  }

  function openAccountForm(kind, options) {
    options = options || {};
    var titles = { bank: "New bank account", cash: "New cash box", overdraft: "New overdraft" };
    var name = el("input", { type: "text", value: options.presetName || "",
      placeholder: kind === "bank" ? "Nabil Bank, current account" : "" });
    var bankName = el("input", { type: "text" });
    var bankAccount = el("input", { type: "text" });
    var bankBranch = el("input", { type: "text" });
    var opening = UI.amountInput("");
    var openingSide = UI.select([
      { value: "dr", label: kind === "overdraft" ? "Debit, in credit" : "Debit, money held" },
      { value: "cr", label: kind === "overdraft" ? "Credit, money owed" : "Credit, overdrawn" }
    ], kind === "overdraft" ? "cr" : "dr");
    var notes = el("textarea", { rows: "2" });

    var bankBlock = el("div", {}, [
      el("div.row", {}, [
        UI.field("Bank name", bankName),
        UI.field("Account number", bankAccount),
        UI.field("Branch", bankBranch)
      ])
    ]);
    if (kind === "cash") { bankBlock.style.display = "none"; }

    UI.modal(titles[kind] || "New account", el("div", {}, [
      el("p.card-note", { text: kind === "cash"
        ? "A cash box is money physically held, such as the counter float or petty cash."
        : kind === "overdraft"
        ? "An overdraft is money the bank has lent, so it sits among the liabilities rather than the assets."
        : "This creates the ledger and puts it under Bank Balances, ready to reconcile." }),
      UI.field("Name", name),
      bankBlock,
      el("div.row", {}, [
        UI.field("Opening balance", opening, "The balance on the day the books begin"),
        UI.field("Side", openingSide)
      ]),
      UI.field("Notes", notes)
    ]), [
      { label: "Cancel" },
      { label: "Add", kind: "primary", action: function () {
        if (!name.value.trim()) { UI.flash("Give the account a name.", "bad"); return false; }
        return api("/api/banking/accounts/create", { body: {
          name: name.value.trim(), kind: kind,
          bank_name: bankName.value.trim(), bank_account_no: bankAccount.value.trim(),
          bank_branch: bankBranch.value.trim(),
          opening: opening.value || 0, opening_side: openingSide.value,
          notes: notes.value.trim()
        }}).then(function (made) {
          UI.flash(name.value.trim() + " added.", "good");
          if (options.onSaved) {
            return api("/api/accounts/" + made.id).then(function (data) {
              options.onSaved(data.account);
            });
          }
          App.go("banking");
        });
      }}
    ]);
  }

  /* Reconciliation */

  App.register("reconcile", function (page) {
    var accountInput = el("input", { type: "text", placeholder: "Choose a bank account" });
    var accountId = App.state.pendingRecon || null;
    App.state.pendingRecon = null;
    var dateField = UI.dateField(NP.todayIso(), function () { if (accountId) { load(); } });
    var statementBalance = UI.amountInput("", { onChange: function () { redrawSummary(); } });
    var box = el("div", {}, [el("div.empty", {}, [
      el("strong", { text: "Choose an account" }),
      el("span", { text: "Pick the bank account, enter the closing balance from the statement, then tick off what the bank has dealt with." })
    ])]);
    var current = null;

    UI.attachPicker(accountInput, function (term) {
      return api("/api/banking/accounts").then(function (data) {
        return data.rows.filter(function (row) {
          return row.account_kind === "bank"
            && (!term || row.name.toLowerCase().indexOf(term.toLowerCase()) >= 0);
        });
      });
    }, function (account) {
      accountId = account.id;
      accountInput.value = account.name;
      load();
    }, function (account) {
      return { main: account.name, side: UI.rs(account.balance) };
    }, {
      createLabel: "Add bank account",
      onCreate: function (typed, done) {
        openAccountForm("bank", { presetName: typed, onSaved: done });
      }
    });

    function load() {
      return api("/api/banking/reconciliation", { query: {
        account_id: accountId, statement_date_ad: dateField.getIso()
      }}).then(function (data) {
        current = data;
        accountInput.value = data.account.name;
        draw();
      });
    }

    function draw() {
      UI.clear(box);
      var data = current;
      box.appendChild(el("div.grid.two", {}, [summaryCard(), historyCard()]));

      var rows = data.lines.map(function (line) {
        var tick = el("input", { type: "checkbox" });
        tick.checked = line.cleared;
        tick.addEventListener("change", function () {
          var body = { entry_ids: [line.entry_id], account_id: data.account.id,
                       cleared_ad: tick.checked ? dateField.getIso() : "" };
          api("/api/banking/clear", { body: body }).then(function () {
            line.cleared = tick.checked;
            line.cleared_ad = tick.checked ? dateField.getIso() : "";
            recompute();
          }).catch(function (error) {
            tick.checked = !tick.checked;
            UI.flash(error.message, "bad");
          });
        });
        var row = el("tr" + (line.cleared ? ".is-cleared" : ""), {}, [
          el("td.mid", {}, [tick]),
          el("td", { text: UI.bs(line.date_ad, "short") }),
          el("td", {}, [el("button.link-button", { text: line.number,
            onclick: function () { Vouchers.view(line.voucher_id); } })]),
          el("td", { text: line.instrument_no || "" }),
          el("td", { text: line.party_name || line.particulars, style: "max-width:260px" }),
          el("td.num", { text: UI.rs(line.dr, { blankZero: true }) }),
          el("td.num", { text: UI.rs(line.cr, { blankZero: true }) })
        ]);
        line.node = row;
        return row;
      });

      box.appendChild(el("div.card", {}, [
        el("div.card-head", {}, [
          el("h2", { text: "Entries up to the statement date" }),
          el("div.row.no-print", {}, [
            el("button.secondary", { text: "Tick everything", onclick: function () { bulk(true); } }),
            el("button.secondary", { text: "Untick everything", onclick: function () { bulk(false); } })
          ])
        ]),
        el("p.card-note", { text: "Tick a line once it appears on the bank statement. What stays unticked is exactly what explains the difference." }),
        UI.table([
          { label: "Cleared", mid: true }, "Date", "Voucher", "Cheque or reference", "Particulars",
          { label: "Paid in", num: true }, { label: "Paid out", num: true }
        ], rows, null, { tall: true, emptyText: "Nothing has gone through this account yet." })
      ]));
    }

    function bulk(state) {
      var ids = current.lines
        .filter(function (line) { return line.cleared !== state; })
        .map(function (line) { return line.entry_id; });
      if (!ids.length) { return; }
      api("/api/banking/clear", { body: {
        entry_ids: ids, account_id: current.account.id,
        cleared_ad: state ? dateField.getIso() : ""
      }}).then(function () { load(); })
        .catch(function (error) { UI.flash(error.message, "bad"); });
    }

    function recompute() {
      var depositsOut = 0, paymentsOut = 0;
      current.lines.forEach(function (line) {
        if (line.node) { line.node.className = line.cleared ? "is-cleared" : ""; }
        if (!line.cleared) { depositsOut += line.dr; paymentsOut += line.cr; }
      });
      current.uncleared_deposits = depositsOut;
      current.uncleared_payments = paymentsOut;
      current.implied_statement_balance = current.book_balance - depositsOut + paymentsOut;
      current.cleared_count = current.lines.filter(function (l) { return l.cleared; }).length;
      current.uncleared_count = current.lines.length - current.cleared_count;
      redrawSummary();
    }

    /* The working is built once and then only its numbers are changed. An
       earlier version rebuilt it on every keystroke, which pulled the box being
       typed in out of the page underneath the cursor. */

    var summaryNodes = {};
    var summaryBody = el("div.recon-summary");
    var summaryBuilt = false;

    function buildSummary() {
      UI.clear(summaryBody);
      summaryNodes = {};
      summaryNodes.book = line("Balance as per our books");
      summaryNodes.deposits = line("Less deposits the bank has not credited yet");
      summaryNodes.payments = line("Add cheques issued that are not presented yet");
      summaryNodes.implied = line("Balance the statement should show", "rule");
      summaryBody.appendChild(el("div", { style: "margin:.6rem 0 .2rem" }, [
        UI.field("Closing balance on the bank statement", statementBalance)
      ]));
      summaryNodes.difference = line("Still to be explained");
      summaryNodes.differenceRow = summaryNodes.difference.parentNode;

      summaryNodes.saveButton = el("button.secondary", { text: "Save progress",
        onclick: function () { save(false); } });
      summaryNodes.doneButton = el("button.primary", { text: "Mark it reconciled",
        onclick: function () { save(true); } });
      summaryBody.appendChild(el("div.row", { style: "margin-top:.7rem" }, [
        summaryNodes.saveButton, summaryNodes.doneButton,
        el("button.secondary.no-print", { text: "Print", onclick: UI.printPage }),
      UI.exportButton()
      ]));
      summaryNodes.counts = el("p.card-note", { style: "margin-top:.5rem" });
      summaryBody.appendChild(summaryNodes.counts);
      summaryBuilt = true;
    }

    function line(label, kind) {
      var value = el("span.num");
      summaryBody.appendChild(el("div.line" + (kind ? "." + kind : ""), {}, [
        el("span", { text: label }), value
      ]));
      return value;
    }

    function summaryCard() {
      if (!summaryBuilt) { buildSummary(); }
      redrawSummary();
      return el("div.card", {}, [
        el("div.card-head", {}, [el("h2", { text: "The working" })]),
        summaryBody
      ]);
    }

    function redrawSummary() {
      if (!current || !summaryBuilt) { return; }
      summaryNodes.book.textContent = UI.rs(current.book_balance);
      summaryNodes.deposits.textContent = UI.rs(-current.uncleared_deposits);
      summaryNodes.payments.textContent = UI.rs(current.uncleared_payments);
      summaryNodes.implied.textContent = UI.rs(current.implied_statement_balance);

      var typed = String(statementBalance.value || "").trim();
      var stated = NP.toPaisa(typed || 0);
      var difference = stated - current.implied_statement_balance;
      var row = summaryNodes.differenceRow;
      if (!typed) {
        row.style.display = "none";
      } else {
        row.style.display = "";
        row.className = "line " + (difference === 0 ? "match" : "off");
        row.firstChild.textContent = difference === 0 ? "It agrees" : "Still to be explained";
        summaryNodes.difference.textContent = UI.rs(difference);
      }
      summaryNodes.doneButton.disabled = !typed || difference !== 0;
      summaryNodes.counts.textContent = current.cleared_count + " cleared, "
        + current.uncleared_count + " still outstanding.";
    }

    function historyCard() {
      var wrap = el("div.card", {}, [
        el("div.card-head", {}, [el("h2", { text: "Earlier reconciliations" })]),
        el("div.empty", { text: "None yet." })
      ]);
      api("/api/banking/history", { query: { account_id: accountId } }).then(function (data) {
        UI.clear(wrap);
        wrap.appendChild(el("div.card-head", {}, [el("h2", { text: "Earlier reconciliations" })]));
        var rows = data.rows.map(function (row) {
          return el("tr", {}, [
            el("td", { text: UI.bs(row.statement_date_ad, "short") }),
            el("td.num", { text: UI.rs(row.statement_balance) }),
            el("td.num", { text: UI.rs(row.book_balance) }),
            el("td", {}, [el("span.pill" + (row.status === "completed" ? ".good" : ".warn"),
              { text: row.status })]),
            el("td.muted", { text: row.created_by, style: "font-size:.76rem" })
          ]);
        });
        wrap.appendChild(UI.table([
          "Statement date", { label: "Per statement", num: true },
          { label: "Per books", num: true }, "Status", "By"
        ], rows, null, { emptyText: "None yet for this account." }));
      });
      return wrap;
    }

    function save(complete) {
      api("/api/banking/reconciliation/save", { body: {
        account_id: accountId, statement_date_ad: dateField.getIso(),
        statement_balance: statementBalance.value || 0, complete: complete ? 1 : 0
      }}).then(function () {
        UI.flash(complete ? "Reconciled and recorded." : "Progress saved.", "good");
        load();
      }).catch(function (error) { UI.flash(error.message, "bad"); });
    }

    page.appendChild(el("div.card", {}, [
      el("div.card-head", {}, [el("h2", { text: "Bank reconciliation" })]),
      el("div.toolbar", {}, [
        el("div.field", { style: "flex:1 1 250px;margin:0" }, [
          el("label", { text: "Account" }), accountInput
        ]),
        UI.field("Statement date", dateField)
      ])
    ]));
    page.appendChild(box);

    if (accountId) { return load(); }
  });

  return { openAccountForm: openAccountForm };
}());
